Add toEncoding function to ToJSON instance of ShelleyGenesis and Shel…

newhoggy opened the pull request:
…leyGenesisStaking and use strict evaluation to help with lazy streaming

lehins commented on eras/shelley/impl/src/Cardano/Ledger/Shelley/Genesis.hs:
```diff
@@ -167,6 +167,43 @@ instance Era era => ToJSON (ShelleyGenesis era) where
+        !protocolParams = sgProtocolParams sg
+        !genDelegs = sgGenDelegs sg
+        !initialFunds = sgInitialFunds sg
+        !staking = sgStaking sg
```
If you want to ensure `sgInitialFunds` and `sgStaking` are strict then it would make sense to manually bind them with a `!`, but for other fields that are already strict this is just unnecessary boiler plate.  Can you please switch all others to `NamedFieldPuns`

lehins commented on eras/shelley/impl/src/Cardano/Ledger/Shelley/Genesis.hs:
```diff
@@ -199,6 +236,17 @@ instance CC.Crypto crypto => ToJSON (ShelleyGenesisStaking crypto) where
+    -- referenced, which helps prevent retention of any fields that have already
+    -- been serialised.
+    let !pools = sgsPools sgs
+        !stake = sgsStake sgs
```
Comment and strict let bindings are unnecessary, both of these fields are strict. Can you switch them to NamedFieldPuns?

lehins reviewed (dismissed):
Looks good overall, but I think it will be nicer if we can reduce boiler plate and redundant strictness annotations.

lehins commented:
@newhoggy Also I'd like to ask you to improve the commit message a little. We are trying to get better at it in ledger. Here is short guide about it: https://github.com/input-output-hk/cardano-ledger/blob/master/CONTRIBUTING.md#commit-messages

lehins reviewed (approved):
LGTM

newhoggy committed:
Enable lazy streaming when JSON encoding ShelleyGenesis and ShelleyGenesisStaking.

This is done by defining toEncoding which avoids building a Value and carefully applying strictness to ensure parent objects don't hold on memory that have already been encoded.
